PEDS/peds13-2.py

```diff
@@ -0,0 +1,491 @@
+    opened_files_list = openfile_cursor.fetchall()
+
+    #select the list of opened files (just wrote file) of pid
+    written_files_query = '''
+            SELECT process, name, mode
+            FROM opened_files
+            WHERE mode == 2 
+            '''
+    writefile_cursor.execute(written_files_query)
+    total_files = writefile_cursor.fetchall()
+
+    topenedf = [] #total opened files from the matrix file
+    for file in opened_files_list:
+        for line in total_files:
+            if line[1] in file[1]:
+                topenedf.append(file) if file not in topenedf else None
+
+    #select the parent id of pid from process list
+    process_parent_query = '''
+                SELECT parent
+                FROM processes
+                WHERE id = %s
+                '''
+    parent_cursor.execute(process_parent_query % pid)
+    parent_id = parent_cursor.fetchall()
+
+    # add data process of pid to linked list
+    process_node.add(topenedf, pid, parent_id, process_name, -1)
+    # calling again this class recursively for the child of process
+    for child in child_list:
+        if child[0] != None :
+            process_node.append(pid, CreateTree(child[0], process_node,db_path))
+    data = process_node.get_data(pid)
+    return data
+
+def main():
+    parser = argparse.ArgumentParser(description='Classification of the pipeline processes and making its graph')
+    parser.add_argument("-db", "--sqliteDB",
+                        help="The path to the sqlite file which is created by reprozip trace and includes all pipeline process")
+    parser.add_argument("-ofile", "--openedFiles",
+                        help="refers to the matrix file of output of the 'repro-tools' script")
+    args = parser.parse_args()
+    graph = Di('Graph', filename='GraphModel', format='dot', strict=False)
+    node_label = 0
+    proc_list = []
+    db_path=args.sqliteDB
+    read_matrix_file=args.openedFiles
+    # write two output files for processes and files separately
+    write_files = open("complete_file.txt", 'w')
+    write_proc = open("all_processes", 'w')
+    write_total_tmp = ['000']
+    write_total_tmp2 = ['000']
+
+    # read the pipeline files :
+    with open(read_matrix_file, 'r') as pfiles:
+        pipeline_files = pfiles.readlines()
+    # read the whole files
+    db = sqlite3.connect(db_path)
+    writefile_cursor = db.cursor()
+    #select the list of opened files (just written files)
+    written_files_query = '''
+            SELECT process, name, mode
+            FROM opened_files
+            WHERE mode == 2 
+            '''
+    writefile_cursor.execute(written_files_query)
+    written_files_list = writefile_cursor.fetchall()
+    db.close()
+
+    #start the program with the root ptocess id, one:
+    pipeline_graph = LinkedList()
+    CreateTree(1, pipeline_graph,db_path)
+    pipeline_graph.reverse()
+    pipeline_graph.filter()
+    pipeline_graph.reverse()
```

Comment:
after creating the list we need to reverse it to refer the head of the list to the root process. but the second revers is needed because it will be reverse again inside the filtering function and I put it to bring the root process in the head of the graph.